src/server.c

```diff
@@ -5164,47 +5814,94 @@ NULL
         addReplyArrayLen(c, dictSize(server.commands));
         di = dictGetIterator(server.commands);
         while ((de = dictNext(di)) != NULL) {
-            addReplyCommand(c, dictGetVal(de));
+            addReplyCommandLegacy(c, dictGetVal(de));
         }
         dictReleaseIterator(di);
     } else if (!strcasecmp(c->argv[1]->ptr, "info")) {
         int i;
         addReplyArrayLen(c, c->argc-2);
         for (i = 2; i < c->argc; i++) {
-            addReplyCommand(c, dictFetchValue(server.commands, c->argv[i]->ptr));
+            addReplyCommandLegacy(c, lookupCommandBySds(c->argv[i]->ptr));
         }
     } else if (!strcasecmp(c->argv[1]->ptr, "count") && c->argc == 2) {
         addReplyLongLong(c, dictSize(server.commands));
     } else if (!strcasecmp(c->argv[1]->ptr,"getkeys") && c->argc >= 3) {
-        struct redisCommand *cmd = lookupCommand(c->argv[2]->ptr);
-        getKeysResult result = GETKEYS_RESULT_INIT;
-        int j;
+        getKeysSubcommand(c);
+    } else {
+        addReplySubcommandSyntaxError(c);
+    }
+}
 
-        if (!cmd) {
-            addReplyError(c,"Invalid command specified");
-            return;
-        } else if (cmd->getkeys_proc == NULL && cmd->key_specs_num == 0) {
-            addReplyError(c,"The command has no key arguments");
-            return;
-        } else if ((cmd->arity > 0 && cmd->arity != c->argc-2) ||
-                   ((c->argc-2) < -cmd->arity))
-        {
-            addReplyError(c,"Invalid number of arguments specified for command");
-            return;
-        }
+/* COMMANDS COUNT */
+void commandsCountCommand(client *c) {
+    addReplyLongLong(c, dictSize(server.commands));
+}
 
-        if (!getKeysFromCommand(cmd,c->argv+2,c->argc-2,&result)) {
-            addReplyError(c,"Invalid arguments specified for command");
-        } else {
-            addReplyArrayLen(c,result.numkeys);
-            for (j = 0; j < result.numkeys; j++) addReplyBulk(c,c->argv[result.keys[j]+2]);
+/* COMMANDS LIST */
+void commandsListCommand(client *c) {
+    dictIterator *di;
+    dictEntry *de;
+
+    addReplySetLen(c, dictSize(server.commands));
+    di = dictGetIterator(server.commands);
+    while ((de = dictNext(di)) != NULL) {
+        struct redisCommand *cmd = dictGetVal(de);
+        addReplyBulkCString(c,cmd->name);
+    }
+    dictReleaseIterator(di);
+}
+
+/* COMMANDS INFO [<command-name> ...] */
```

Comment:
I think introducing command vs commands is just going to be confusing, since it's going to be easy to forget which is the "new one" and which is the "old one" until we deprecate the old one. I would rather introduce new sub-arguments that introduce the new behavior.

Reply:
the problem with the old COMMAND is that it also works with no arguments (we have no other command with sub-command that does that). so i really wanna keep it entirely unchanged (for backwards compatibility) and encourage people to use the one one.

you're right that COMMAND vs COMMANDS is slightly confusing. maybe we can come up with a different name?

Reply:
Can we not leave the regular `COMMAND` command in place for the old way of listing commands, and introduce the new `COMMAND LIST` sub-command as the new mechanism with the new structure. It seems reasonable to then say that the root command will be deprecated in 7 and clients should start using the new mechanism.

We also might want to think about building a versioning scheme for the Redis command set that clients can set on connect/hello. If they connect with version 1 they can issue COMMAND, but in version 2 it will throw an error since it will be deprecated.

Reply:
So in your suggestion drop that support in 7, old client libraries won't work with 7, and users that want to upgrade redis (to enjoy an optimization or a new feature) won't necessarily be able to do so. 

The other major issue to keep in mind is that client libraries won't be able to just stop using the old COMMAND anytime soon, since they still need to support old redis versions.
 
I discussed this with Guy and we concluded that it's OK to ask them to first issue the new COMMANDS and only if it fails issue the old one. It means an additional round trip on connection. 

I don't like to have redis explicitly support multiple versions old protocols and command sets.. Currently we support all (never remove a command or a protocol), but even if we someday remove something, we can remove them one by one separately, no sense to bind them IMHO. And I don't see a reason to refuse something that's still actually supported if the client is using another new feature. 

Maybe we need to discuss this in more detail in another forum?

Reply:
I'm not saying to drop support in 7, just to deprecate it so that we can remove it in some future version. I don't see why we can't have `COMMAND` and `COMMAND LIST` live together and have a different output?

Reply:
@madolson the main problem is `COMMAND INFO` - we broke its structure, mostly in order to remove the old first,last,step scheme (we were under the impression that as long as the "legacy" scheme exists, people won't put the effort in parsing the new key specs) - @oranagra CMIIW

i don't mind dropping COMMANDS if we can get COMMAND to:
1. output a list of all commands (just a list of strings), with the ability to filer them (for example, get only commands that belong to some module, by regex, etc.)
2. output detailed information about all commands (each such information is not backward compatible with COMMAND INFO, so we can't just add stuff to COMMAND INFO's output). it should also be able to emit information about specific commands.

the most "ugly" thing is that, no matter what, if we choose to use COMMAND, there will be two subcommands that retrieve detailed information about commands: INFO (the old struct) and INFO2/DETAILS/LIST/whatever that use the new struct, with key specs and subcommands

Reply:
In theory, there's no problem to add key specs to COMMAND INFO (same as we have in unstable today), also also add the sub-commands and additional Metadata to that array (old clients will ignore the additional entries in the response). 

So in that sense we can simply add COMMAND LIST (to just list command names) and extend COMMAND INFO, and do nothing else. 

But there are so many issues with the old COMMAND that I rather just leave it unchanged and add all the new features to a new command. 

The issues that bother me:
1. The fact that the command have sub-commands, but also functions as a command with no arguments too. This causes complications to the command and sub-commands listing. We may need to re-introduce the `container` flag to distinguish between and command with optional sub-commands, and one that's just a container for sub-commands. 
2. The fact the response for the command Metadata is array with specific indexes rather than a map with field names. (I don't like to add the `since`, `returns` or other fields we plan in #9359 to just sit in some specific index without a title) 
3. Some old fields that I wish to remove (firstkey etc)

Reply:
@madolson there are two new functionalities:
1. `COMMANDS LIST` - returns a set of command names
2. `COMMANDS INFO` - returns detailed information about commands (like `COMMAND INFO`, but returns a map instead of a list, removes the first,last,step fields, and adds key_specs and subcommand fields)

IIUC you suggest the following alternatives:
1. `COMMAND LIST`
2. `COMMAND INFO2/DETAILS/whatever`

(CMIIW)

@yossigo @soloestoy any thoughts?

Reply:
I edited Guy's post with a few clarifications.
the 3rd options is:

3 . add the new `COMMANDS INFO` features we wanted to add to the existing `COMMAND` (all commands) and `COMMAND INFO <cmd>` (one command) as new fields, without modifying the content of old ones.

what i don't like about that is:
- we still support COMMAND with no argument
- we still use an array of fields in specific indexes (e.g. the sub-commands will be in [9] but will not have any title to it, and people will have to just be aware that [9] is the subcommand list (and [8] is the key-specs

the upside is that client libraries will still be able to make a single COMMAND call on startup and decide how to parse it by the number of fields they see, instead of issuing two commands.
but i actually rather make it more explicit (i think that a completely different command will make that statement)

my preference is for a completely new command (a admit that COMMANDS is a confusing name).
my second choice would be option 2 from the list in the previous post.

Reply:
To me, making this accessible and practical to clients beats the other arguments. In theory we could piggy back on top of `HELLO` so clients can negotiate this, but I think this would be somewhat over-engineered.

Reply:
@guybe7 we discussed this with the core team, we concluded that for now we're gonna just append more elements to the old COMMAND response, so a client can just keep issuing the old COMMAND and get both old and new metadata depending on the redis version.

maybe when we'll handle the next step (commands.json), we'll decide to either add a new command or a new sub-command that has that same information in a map with field names rather than array with predefined indexes.
i.e. the information will be included in the old format, and maybe **also** in a new one, and that'll be decided later.

i suppose we can keep the new COMMAND LIST feature with the same design we planned for COMMANDS LIST